### PR TITLE
Added accuracy control quantization

### PR DESCRIPTION
**Pull request by adrianboguszewski**
## 📝 Description

- Accuracy control quantization in NNCF while exporting to OpenVINO

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/engine/engine.py [hunk 1]
```diff
@@ -891,6 +893,9 @@ def export(
             datamodule (AnomalibDataModule | None, optional): Lightning datamodule.
                 Must be provided if CompressionType.INT8_PTQ is selected.
                 Defaults to ``None``.
+            metric (Metric | str | None, optional): Metric to measure quality loss when quantizing.
```
Comment (ashwinvaidya17) [inline src/anomalib/engine/engine.py:896]: Maybe we should also mention that this is only applicable for OpenVINO export

File: src/anomalib/models/components/base/export_mixin.py [hunk 2]
```diff
@@ -235,20 +239,57 @@ def to_openvino(
...
+                    datamodule.train_batch_size = model_input.shape[0]
+                dataloader = datamodule.train_dataloader()
                 if len(dataloader.dataset) < 300:
                     logger.warning(
                         f">300 images recommended for INT8 quantization, found only {len(dataloader.dataset)} images",
                     )
+
                 calibration_dataset = nncf.Dataset(dataloader, lambda x: x["image"])
                 model = nncf.quantize(model, calibration_dataset)
+            elif compression_type == CompressionType.INT8_ACQ:
```
Comment (ashwinvaidya17) [inline src/anomalib/models/components/base/export_mixin.py:261]: Can we extract nncf related code to another method? Right now it bloats the ov export method.

File: src/anomalib/models/components/base/export_mixin.py [hunk 3]
```diff
@@ -235,20 +239,57 @@ def to_openvino(
...
+                    )
+
+                calibration_dataset = nncf.Dataset(dataloader, lambda x: x["image"])
+                validation_dataset = nncf.Dataset(datamodule.val_dataloader())
+
+                if isinstance(metric, str):
+                    metric = create_metric_collection([metric])[metric]
+
+                # validation function to evaluate the quality loss after quantization
+                def val_fn(nncf_model: ov.CompiledModel, validation_data: Iterable) -> float:
```
Comment (ashwinvaidya17) [inline src/anomalib/models/components/base/export_mixin.py:285]: Is it possible to avoid inline method?

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]
Thanks! Just a few comments

File: src/anomalib/models/components/base/export_mixin.py (hunk 3 above)
Comment (adrianboguszewski) [reply src/anomalib/models/components/base/export_mixin.py:285]: It looks like not possible. I need to use metric and task variables, which cannot be passed to this function as the signature is defined by NNCF.

File: src/anomalib/deploy/export.py [hunk 4]
```diff
@@ -57,6 +57,12 @@ class CompressionType(str, Enum):
     Full integer post-training quantization (INT8)
     All weights and operations are quantized to INT8. Inference is done in INT8 precision.
     """
+    INT8_ACQ = "int8_acq"
+    """
+    Accuracy-control quantization (INT8)
+    Weights and operations are quantized to INT8, except those that would degrade quality of the model more than is
+    acceptable. Inference is done in a mixed precision.
+    """
```
Comment (samet-akcay) [inline src/anomalib/deploy/export.py:65]: Would it be possible to add all of the above commented sections to the docstring. Something like this:
```python
class CompressionType(str, Enum):
    """Model compression type when exporting to OpenVINO.

    Attributes:
        FP16 (str): Weight compression (FP16). All weights are converted to FP16.
        INT8 (str): Weight compression (INT8). All weights are quantized to INT8,
            but are dequantized to floating point before inference.
        INT8_PTQ (str): Full integer post-training quantization (INT8).
            All weights and operations are quantized to INT8. Inference is done
            in INT8 precision.
        INT8_ACQ (str): Accuracy-control quantization (INT8). Weights and
            operations are quantized to INT8, except those that would degrade
            quality of the model more than is acceptable. Inference is done in
            a mixed precision.

    Examples:
        >>> from anomalib.deploy import CompressionType
        >>> CompressionType.INT8_PTQ
        'int8_ptq'
    """

    FP16 = "fp16"
    INT8 = "int8"
    INT8_PTQ = "int8_ptq"
```

This is especially useful to add these to the documentation.

File: src/anomalib/models/components/base/export_mixin.py [hunk 5]
```diff
@@ -176,6 +184,10 @@ def to_openvino(
             datamodule (AnomalibDataModule | None, optional): Lightning datamodule.
                 Must be provided if CompressionType.INT8_PTQ is selected.
                 Defaults to ``None``.
+            metric (Metric | str | None, optional): Metric to measure quality loss when quantizing.
+                Must be provided if CompressionType.INT8_ACQ is selected and must return higher value for better
```
Comment (samet-akcay) [inline src/anomalib/models/components/base/export_mixin.py:188]:
```suggestion
                Must be provided if ``CompressionType.INT8_ACQ`` is selected and must return higher value for better
```

File: src/anomalib/models/components/base/export_mixin.py [hunk 6]
```diff
@@ -257,6 +252,95 @@ def to_openvino(
 
         return ov_model_path
 
+    def _compress_ov_model(
```
Comment (samet-akcay) [inline src/anomalib/models/components/base/export_mixin.py:255]: This function seems to be a bit long. Would it be an idea to split it to improve the readability? Something like:

```python
def _compress_ov_model(...)
        ...
        if compression_type == CompressionType.INT8:
            model = nncf.compress_weights(model)
        elif compression_type == CompressionType.INT8_PTQ:
            model = _post_training_quantization(model, datamodule)
        elif compression_type == CompressionType.INT8_ACQ:
            model = _accuracy_control_quantization(model, datamodule, metric, task)
        else:
            raise ValueError("...")
        return model

```

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks for creating this PR, and for your patience waiting for this to be reviewed. I've got few minor comments.

File: src/anomalib/models/components/base/export_mixin.py [hunk 7]
```diff
@@ -206,6 +218,22 @@ def to_openvino(
...
 
+            Export and Quantize the Model (OpenVINO IR):
+            This example demonstrates how to export and quantize the model to OpenVINO IR.
+
+            >>> from anomalib.models import Patchcore
+            >>> from anomalib.data import Visa
+            ...
+            >>> datamodule = Visa()
+            >>> model = Patchcore()
+            ...
```
Comment (samet-akcay) [inline src/anomalib/models/components/base/export_mixin.py:229]:
```suggestion
            
```

File: src/anomalib/models/components/base/export_mixin.py [hunk 8]
```diff
@@ -206,6 +218,22 @@ def to_openvino(
             ...     task=datamodule.test_data.task
             ... )
 
+            Export and Quantize the Model (OpenVINO IR):
+            This example demonstrates how to export and quantize the model to OpenVINO IR.
+
+            >>> from anomalib.models import Patchcore
+            >>> from anomalib.data import Visa
+            ...
```
Comment (samet-akcay) [inline src/anomalib/models/components/base/export_mixin.py:226]:
```suggestion
            
```

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks, just two super minor comment, then we could merge.

**Review by samet-akcay — APPROVED** [review]

File: src/anomalib/models/components/base/export_mixin.py [hunk 9]
```diff
@@ -235,20 +257,8 @@ def to_openvino(
...
-                    raise ValueError(msg)
-
-                dataloader = datamodule.val_dataloader()
-                if len(dataloader.dataset) < 300:
-                    logger.warning(
-                        f">300 images recommended for INT8 quantization, found only {len(dataloader.dataset)} images",
-                    )
-                calibration_dataset = nncf.Dataset(dataloader, lambda x: x["image"])
-                model = nncf.quantize(model, calibration_dataset)
+            if compression_type != CompressionType.FP16:
```
Comment (adrianboguszewski) [inline src/anomalib/models/components/base/export_mixin.py:260]:
```suggestion
            if compression_type and compression_type != CompressionType.FP16:
```

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks

File: src/anomalib/engine/engine.py [hunk 10]
```diff
@@ -871,6 +872,7 @@ def export(
         transform: Transform | None = None,
         compression_type: CompressionType | None = None,
         datamodule: AnomalibDataModule | None = None,
+        metric: Metric | str | None = None,
```
[the remark this answers is not in the document]
Comment (adrianboguszewski) [reply src/anomalib/engine/engine.py:875]: Ok

**Pull request merged**